pkg/log/log.go

```diff
@@ -0,0 +1,201 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package log
+
+import (
+    "bytes"
+    "fmt"
+    "path"
+    "runtime"
+    "sort"
+    "strings"
+
+    log "github.com/sirupsen/logrus"
+    "gopkg.in/natefinch/lumberjack.v2"
+)
+
+const (
+    defaultLogLevel      = log.InfoLevel
+    defaultLogTimeFormat = "2006/01/02 15:04:05.000"
+)
+
+func Info(v ...interface{}) {
+    log.Info(v...)
+}
+
+func Infof(format string, v ...interface{}) {
+    log.Infof(format, v...)
+}
+
+func Debug(v ...interface{}) {
+    log.Debug(v...)
+}
+
+func Debugf(format string, v ...interface{}) {
+    log.Debugf(format, v...)
+}
+
+func Warn(v ...interface{}) {
+    log.Warn(v...)
+}
+
+func Warnf(format string, v ...interface{}) {
+    log.Warnf(format, v...)
+}
+
+func Error(v ...interface{}) {
+    log.Error(v...)
+}
+
+func Errorf(format string, v ...interface{}) {
+    log.Errorf(format, v...)
+}
+
+func Panic(args ...interface{}) {
+    log.Panic(args...)
+}
+
+func Fatal(v ...interface{}) {
+    log.Fatal(v...)
+}
+
+func Fatalf(format string, v ...interface{}) {
+    log.Fatalf(format, v...)
+}
+
+func stringToLogLevel(level string) log.Level {
+    switch strings.ToLower(level) {
+    case "fatal":
+        return log.FatalLevel
+    case "error":
+        return log.ErrorLevel
+    case "warn", "warning":
+        return log.WarnLevel
+    case "debug":
+        return log.DebugLevel
+    case "info":
+        return log.InfoLevel
+    }
+    return defaultLogLevel
+}
+
+// hook injects file name and line pos into log entry.
+type hook struct{}
+
+// Fire implements logrus.Hook interface
+func (h *hook) Fire(entry *log.Entry) error {
+    // these two num are set by manually testing
+    pc := make([]uintptr, 4)
+    cnt := runtime.Callers(10, pc)
+
+    for i := 0; i < cnt; i++ {
+        fu := runtime.FuncForPC(pc[i] - 1)
+        name := fu.Name()
+        if !isSkippedPackageName(name) {
+            file, line := fu.FileLine(pc[i] - 1)
+            entry.Data["file"] = path.Base(file)
+            entry.Data["line"] = line
+            break
+        }
+    }
+    return nil
+}
+
+// Levels implements logrus.Hook interface.
+func (h *hook) Levels() []log.Level {
+    return log.AllLevels
+}
+
+// isSKippedPackageName tests wether path name is on log library calling stack.
+func isSkippedPackageName(name string) bool {
+    return strings.Contains(name, "github.com/sirupsen/logrus") ||
+            strings.Contains(name, "github.com/apache/pulsar-client-go/pkg/log")
+}
+
+// formatter is for compatibility with ngaut/log
+type formatter struct {
+    DisableTimestamp, EnableEntryOrder bool
+}
+
+// Format implements logrus.Formatter
+func (f *formatter) Format(entry *log.Entry) ([]byte, error) {
```

Comment:
I don't see why do we need to re-implement the logger API. Logrus seems to be well established and the API has been in use for a while. Why not just use that instead of reinventing it?

Reply:
The default form of `logrus` does not provide the line number information of the file. When the log is output, the user may prefer to locate the current log output by the line number information.